compiler/ir/ir.psi2ir/src/org/jetbrains/kotlin/psi2ir/generators/FragmentClassGenerator.kt

```diff
@@ -0,0 +1,70 @@
+/*
+ * Copyright 2010-2021 JetBrains s.r.o. and Kotlin Programming Language contributors.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
+ */
+
+package org.jetbrains.kotlin.psi2ir.generators
+
+import org.jetbrains.kotlin.descriptors.Modality
+import org.jetbrains.kotlin.ir.UNDEFINED_OFFSET
+import org.jetbrains.kotlin.ir.declarations.IrClass
+import org.jetbrains.kotlin.ir.declarations.IrDeclarationOrigin
+import org.jetbrains.kotlin.ir.declarations.addMember
+import org.jetbrains.kotlin.ir.symbols.impl.IrConstructorPublicSymbolImpl
+import org.jetbrains.kotlin.ir.util.createIrClassFromDescriptor
+import org.jetbrains.kotlin.ir.util.defaultType
+import org.jetbrains.kotlin.name.Name
+import org.jetbrains.kotlin.psi.KtBlockCodeFragment
+
+class FragmentClassGenerator(
+    declarationGenerator: FragmentDeclarationGenerator,
+    private val fragmentInfo: EvaluatorFragmentInfo
+) : FragmentDeclarationGeneratorExtension(declarationGenerator) {
+
+    fun generateClassForCodeFragment(ktFile: KtBlockCodeFragment): IrClass {
+        val classDescriptor = fragmentInfo.classDescriptor
+        val startOffset = UNDEFINED_OFFSET
+        val endOffset = UNDEFINED_OFFSET
+        val visibility = classDescriptor.visibility
+        val modality = Modality.FINAL
+
+        return context.symbolTable.declareClass(classDescriptor) {
+            context.irFactory.createIrClassFromDescriptor(
+                startOffset, endOffset, IrDeclarationOrigin.DEFINED, it, classDescriptor,
+                context.symbolTable.nameProvider.nameForDeclaration(classDescriptor), visibility, modality
+            )
+        }.buildWithScope { irClass ->
+            irClass.thisReceiver = context.symbolTable.declareValueParameter(
```

Comment:
Maybe just `irClass.createImplicitParameterDeclarationWithWrappedDescriptor()`

Reply:
This does not appear to be in "scope" of psi2ir. Forgive me, but you must be referring to `compiler/ir/backend.common/src/org/jetbrains/kotlin/backend/common/ir/IrUtils.kt:createImplicitParameterDeclarationWithWrappedDescriptor`?

 As far as I can tell, that resides in `:compiler:backend-common`, already imported by psi2ir.

I must just be holding it wrong :sweat_smile:

Reply:
Oh, indeed, sorry!

Reply:
Can you help me understand why it's not available? Seems to me it ought to be given the current build configuration.

Reply:
It's declared in the module `:compiler:ir.backend.common`, which `:compiler:ir.psi2ir` doesn't depend on